### PR TITLE
Fixed duplicate creation of startup actors on reconnection

### PR DESCRIPTION
**Pull request by Vatyx**
#### Description
On reconnect for a server, even if startup actors already have entities, they will not be created again, nor will BeginPlay be called authoritatively again.

However, on reconnection, those entities will be moved back to where they are on level startup rather than where they are in SpatialOS and that needs to be fixed.

Engine PR: https://github.com/improbableio/UnrealEngine/pull/94

#### Release note
Bugfix: Fixed duplicate creation of startup actors on reconnection

#### Tests
Tested them in a local, simple, map I have. Should be tested more.

#### Primary reviewers
@improbable-valentyn @m-samiec

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp [hunk 1]
```diff
@@ -150,7 +150,7 @@ Worker_RequestId USpatialSender::CreateEntity(USpatialActorChannel* Channel)
 	// We use this to indicate if a new Actor should be created or to link a pre-existing Actor
 	// when receiving an AddEntityOp.
 	TSchemaOption<FUnrealObjectRef> StablyNamedObjectRef;
-	if (Actor->IsFullNameStableForNetworking())
+	if (Actor->HasAnyFlags(RF_WasLoaded))
```
Comment (BenNaccarato) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp]: Why was this changed? Are the other conditions in `IsFullNameStableForNetworking` not relevant?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 2]
```diff
@@ -658,24 +627,70 @@ void USpatialReceiver::CleanupDeletedEntity(Worker_EntityId EntityId)
...
+		}
+	}
+
+	// Otherwise, this Actor already exists in the map, get it from the package map.
+	const FUnrealObjectRef& StablyNamedRef = UnrealMetadata->StablyNamedRef.GetValue();
+	AActor* StaticActor = Cast<AActor>(PackageMap->GetObjectFromUnrealObjectRef(StablyNamedRef));
+	// An unintended side effect of GetObjectFromUnrealObjectRef is that this ref
+	// will be registered with this Actor. It can be the case that this Actor is not
+	// stably named (due to bNetLoadOnCliwent = false) so we should let
+	// SpatialPackageMapClient::ResolveEntityActor to handle it properly.
```
Comment (mattyoung-improbable) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]:
```suggestion
	// SpatialPackageMapClient::ResolveEntityActor handle it properly.
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 3]
```diff
@@ -658,24 +627,70 @@ void USpatialReceiver::CleanupDeletedEntity(Worker_EntityId EntityId)
...
+	if (ActorClass == nullptr)
 	{
-		//bRemoteOwned needs to be public in source code. This might be a controversial change.
-		FActorSpawnParameters SpawnInfo;
-		SpawnInfo.SpawnCollisionHandlingOverride = ESpawnActorCollisionHandlingMethod::AlwaysSpawn;
-		SpawnInfo.bRemoteOwned = !NetDriver->IsServer();
-		SpawnInfo.bNoFail = true;
-		// We defer the construction in the GDK pipeline to allow initialization of replicated properties first.
-		SpawnInfo.bDeferConstruction = bDeferred;
+		return NewActor;
```
[the remark this answers is not in the document]
Comment (Vatyx) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: Yeah ok ben SHEESH can you calm down with the hate.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 4]
```diff
@@ -394,22 +393,13 @@ void USpatialReceiver::ReceiveActor(Worker_EntityId EntityId)
...
-				if (EntityActor == nullptr)
-				{
-					// In native networking, if Unreal tries to look up a stably named actor on the client
-					// and it doesn't exist (e.g. streaming level hasn't loaded in) Unreal seems to not do anything.
-					// returning here does the same behavior.
-					return;
-				}
+				// In native networking, if Unreal tries to look up a stably named actor on the client
+				// and it doesn't exist (e.g. streaming level hasn't loaded in) Unreal seems to not do anything.
+				// returning here does the same behavior.
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]:
```suggestion
				// Returning here does the same behavior.
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp (hunk 1 above)
Comment (improbable-valentyn) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp]: `IsFullNameStableForNetworking` relies on `bNetStartup`, which is only going to be set if `bNetLoadOnClient` is true, while we're interested in all the loaded actors.
@Vatyx should probably add a comment.

Comment (Vatyx) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp]: Can do.
